docs/architecture/014-userids.md

```diff
@@ -0,0 +1,192 @@
+# 014: User identities
+
+Authors: @opqdonut @Macroz @jaakkocsc
+
+# Background
+
+Currently REMS uses the userid received from the OIDC server (from the
+`sub` claim or another claim based on configuration), and uses this
+userid as the internal database key for that user. The userid is
+embedded in application events, workflows, etc.
+
+## Problems with the current approach
+
+- The `sub` attribute might be opaque and not usable between systems, e.g. when sending entitlements to other systems
+- Different login methods may have different formats for external user identifiers
+- Different login methods might use conflicting values for claims
+  - E.g. `foo@csc.fi` via HAKA or CSC SSO (TODO: do we know of cases like this?)
+- A single user might have multiple external identifiers (TODO: do we know of cases like this?)
+
+## Use cases for user identities
+
+A user identifier can be used for multiple things. We might want to
+use different identifiers for different purposes.
+
+- Referring to users inside REMS data (e.g. workflow handler)
+- Referring to users in API responses (e.g. workflow handler)
+  - enriching
+- Referring to users in API requests (e.g. set workflow handler, get entitlements for user)
+- Referring to users in entitlement POST calls
+
+# Decision: internal user identifiers
+
+We'll add internal random user ids to REMS. This internal user id will
+be the key that users are referred to within REMS. The user's external
+id will be stored in the attributes JSON blob.
+
+This will allow us more flexibility in the future when identity
+requirements and use cases change, and might also make all sorts of
+migrations easier (since internal user ids don't need to be touched).
+
+## Login
+
+When a user logs in, REMS fetches the user's id token.
+
+1. Find the first attribute from `:oidc-user-attributes` that is set in the id token. Call this attribute-value pair `attr: value`.
+2. Find a user from the REMS database that has `attr: value` stored in their user attributes
+   - If a user was found, set current session's identity to the internal user id of that user.
+   - If a user was not found
+     - Create a new internal user id and store `attr: value` in its attributes.
+     - Set the current session's identity to the just created internal user id
+
+## User identity resolution
+
+When somebody uses the userid `xyz` in an API call, REMS searches for the user like this:
+
+1. If there is a user with a REMS internal user id `xyz`, use that user
+2. Iterate through every attribute `attr` in `:oidc-user-attributes`, in order
+   1. If there is a user with `attr: xyz`, use that user
+   2. If not, keep iterating
+3. If no user was found, return HTTP 400 Bad Request
+
```

Comment:
Should we also support search with `"attr: xyz"` so that you can use any attribute for the user?

Reply:
I'll write that up as a future option.

Reply:
Ah that's actually already mentioned in Open questions > API format

Reply:
Yeah, was just thinking that it should be as a string here but it's commented also later.